iis/datadog_checks/iis/iis.py

```diff
@@ -55,62 +55,71 @@ def get_iishost(self, instance):
+        sites = instance.get('sites', [])
+        expected_sites = set(sites)
         # _Total should always be in the list of expected sites; we always
         # report _Total
         expected_sites.add(TOTAL_SITE)
 
         self.log.debug("Expected sites is {}".format(expected_sites))
-        key = hash_mutable(instance)
-        for inst_name, dd_name, metric_func, counter in self._metrics[key]:
+        instance_hash = hash_mutable(instance)
+        for inst_name, dd_name, metric_func, counter in self._metrics[instance_hash]:
             try:
-                try:
-                    vals = counter.get_all_values()
-                except Exception as e:
-                    self.log.error("Failed to get_all_values {} {}: {}".format(inst_name, dd_name, e))
-                    continue
-
-                for sitename, val in iteritems(vals):
-                    tags = []
-                    if key in self._tags:
-                        tags = list(self._tags[key])
-                    tags.append(self.get_iishost(instance))
-
-                    try:
-                        if not counter.is_single_instance():
-                            if sites and sitename != TOTAL_SITE and sitename not in sites:
-                                continue
-                            tags.append("site:{}".format(self.normalize(sitename)))
-                    except Exception as e:
-                        self.log.error("Caught exception {} setting tags".format(e))
+                site_values = counter.get_all_values()
+            except Exception as e:
+                self.log.error("Failed to get_all_values {} {}: {}".format(inst_name, dd_name, e))
+                continue
+            try:
+                for site_name, site_uptime in iteritems(site_values):
+                    if not counter.is_single_instance():
+                        if site_name != TOTAL_SITE and site_name not in sites:
+                            continue
+                    tags = self._get_site_tags(instance_hash, instance, site_name)
 
                     try:
-                        metric_func(dd_name, val, tags)
+                        metric_func(dd_name, site_uptime, tags)
                     except Exception as e:
-                        self.log.error("Error in metric_func: {} {} {}".format(dd_name, val, e))
+                        self.log.error("Error in metric_func: {} {} {}".format(dd_name, site_uptime, e))
 
                     if dd_name == "iis.uptime":
-                        uptime = int(val)
-                        status = AgentCheck.CRITICAL if uptime == 0 else AgentCheck.OK
-                        self.service_check(self.SERVICE_CHECK, status, tags)
-                        if sitename in expected_sites:
-                            self.log.debug("Removing {!r} from expected sites".format(sitename))
-                            expected_sites.remove(sitename)
+                        self._report_uptime(site_uptime, tags)
```

Comment:
Not exactly sure I see the point in moving code into a function called once.

Reply:
Readability